### ISSUE TITLE
CPU limits don't work (CFS gets disabled) on CentOS7 (3.10.0-x kernel)

### ISSUE THREAD

**Issue by tallaxes** [rancher/rke2#689]
On CentOS7 (cgroups v1) RKE2 incorrectly disables CFS, starting kubelet with `--cpu-cfs-quota=false`. (The message "... level=warning msg="Disabling CPU quotas due to missing cpu.cfs_period_us" is observed in the RKE2 log; issued [here](https://github.com/k3s-io/k3s/blob/c5e2676d5c43aa85ed1e5ac900235f3ce7cf6a1f/pkg/daemons/agent/agent.go#L132-L136)). As a result kubelet does not configure CPU limit enforcement. This likely also affects K3S, and other 3.10.x Kernel -based distributions, but I have only tested with RKE2 (v1.19.7+rke2r1) / CentOS7 (3.10.0-1127.19.1.el7.x86_64).

The underlying problem is likely in K3S (where it has been [diagnosed](https://github.com/k3s-io/k3s/issues/392#issuecomment-553218339) once, but not yet fixed). K3S/RKE2 [checkCgroups](https://github.com/k3s-io/k3s/blob/732228fd159acbe9db2af849df24699317fee7d6/pkg/daemons/agent/agent.go#L214-L228) logic in the agent, responsible for determining CFS capability, expects the relevant entry in `/proc/{pid}/cgroup` file to have systems part with `cpu` (e.g. `cpu,cpuacct`) that matches the name of folder in `/sys/fs/cgroup`. This is not the case in CentOS7 (and possibly other 3.10.0 -based distros?), where `/proc/{pid}/cgroup` files have `cpuacct,cpu` entry, but `/sys/fs/cgroup` has `cpu,cpuacct` folder (order reversed).

The proper fix would be to update the logic (and as far as I can tell, the fix proposed in https://github.com/k3s-io/k3s/pull/1057 would work), but the following workaround (as root) works as well, by adding `cpuacct,cpu` symbolic link (that K3S/RKE2 processes correctly) to `/sys/fs/cgroup` folder. (This likely needs to be done prior to starting K3S/RKE2; restarting RKE2 after the fact will properly start kubelet without `--cpu-cfs-quota=false` flag, but the already running containers won't have their CFS quota set correctly ...)

```sh
# create /sys/fs/cgroup/cpuacct,cpu symlink (workaround for K3S/RKE2 to detect CFS)
mount -o remount,rw /sys/fs/cgroup
ln -s cpu,cpuacct /sys/fs/cgroup/cpuacct,cpu
mount -o remount,ro /sys/fs/cgroup
```

**Comment by brandond** [comment]
I'm on EL7 and I have:
```
[sysadm@centos01 ~]$ cat /etc/redhat-release
CentOS Linux release 7.9.2009 (Core)
[sysadm@centos01 ~]$ grep cpu /proc/self/cgroup
9:cpu,cpuacct:/
7:cpuset:/
[sysadm@centos01 ~]$ ls -lad /sys/fs/cgroup/*cpu*
lrwxrwxrwx. 1 root root 11 Jan 22 01:36 /sys/fs/cgroup/cpu -> cpu,cpuacct
lrwxrwxrwx. 1 root root 11 Jan 22 01:36 /sys/fs/cgroup/cpuacct -> cpu,cpuacct
dr-xr-xr-x. 5 root root  0 Jan 22 01:36 /sys/fs/cgroup/cpu,cpuacct
dr-xr-xr-x. 4 root root  0 Jan 22 01:36 /sys/fs/cgroup/cpuset
```

I am guessing that the order is not deterministic, and that we shouldn't really rely on either. As you can see from the symlinks they are all just aliases for the same thing.

**Comment by tallaxes** [comment]
```
# cat /etc/redhat-release 
CentOS Linux release 7.8.2003 (Core)
# grep cpu /proc/self/cgroup 
10:cpuset:/
3:cpuacct,cpu:/user.slice
# ls -lad /sys/fs/cgroup/*cpu*
lrwxrwxrwx. 1 root root 11 Feb  3 23:05 /sys/fs/cgroup/cpu -> cpu,cpuacct
lrwxrwxrwx. 1 root root 11 Feb  3 23:05 /sys/fs/cgroup/cpuacct -> cpu,cpuacct
drwxr-xr-x. 5 root root  0 Feb  3 23:05 /sys/fs/cgroup/cpu,cpuacct
drwxr-xr-x. 3 root root  0 Feb  3 23:05 /sys/fs/cgroup/cpuset
```

I wonder if 7.8 vs 7.9 makes a difference ...

I have not looked into what determines the order in /proc/{pid}/cgroup, but it looks like the "alternative" order has been observed more than once. I am guessing the links under /sys/fs/cgroup are more consistent. In any case, you are likely right that the logic should not depend on the order.

**Comment by brandond** [comment]
This should be fixed on K3s now, will need a pull-through into RKE2 before it's fixed over here. Probably will happen in the 1.20.3 timeline.

**Comment by ShylajaDevadiga** [comment]
Validated the fix using rke2 version v1.20.4-rc2+rke2r1

Order seems to be same in rke2 verssion v1.19.8. 
Quota is not disabled after the fix as seen in logs of v1.19.8
With cpu resource limit set to "1", limit is honored with quota enabled.

rke2 version v1.20.4-rc2+rke2r1
```
[centos@ip-172-31-11-130 ~]$ rke2 -v
rke2 version v1.20.4-rc2+rke2r1 (10d40f8e604f4ebdc196206b40cb7ec1d2be4358)
go version go1.15.8b5

[centos@ip-172-31-11-130 ~]$ cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)

[centos@ip-172-31-11-130 ~]$ grep cpu /proc/self/cgroup
6:cpuacct,cpu:/user.slice
5:cpuset:/

[centos@ip-172-31-11-130 ~]$ ls -lad /sys/fs/cgroup/*cpu*
lrwxrwxrwx. 1 root root 11 Mar  4 16:41 /sys/fs/cgroup/cpu -> cpu,cpuacct
lrwxrwxrwx. 1 root root 11 Mar  4 16:41 /sys/fs/cgroup/cpuacct -> cpu,cpuacct
drwxr-xr-x. 5 root root  0 Mar  4 16:41 /sys/fs/cgroup/cpu,cpuacct
drwxr-xr-x. 3 root root  0 Mar  4 16:41 /sys/fs/cgroup/cpuset

[centos@ip-172-31-11-130 ~]$ journalctl -u rke2-server |grep cpu
[centos@ip-172-31-11-130 ~]$

[centos@ip-172-31-11-130 ~]$ k top pods cpulimit
NAME       CPU(cores)   MEMORY(bytes)
cpu-demo   1000m        0Mi
[centos@ip-172-31-11-130 ~]$

```
rke2 version v1.19.8+rke2r1

```
[centos@ip-172-31-1-193 ~]$ rke2 -v
rke2 version v1.19.8+rke2r1 (77ad545580efc4ab64ce231d4a75fa2e493956a3)
go version go1.15.8b5

[centos@ip-172-31-1-193 ~]$ cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)

[centos@ip-172-31-1-193 ~]$ grep cpu /proc/self/cgroup
9:cpuacct,cpu:/user.slice
5:cpuset:/

[centos@ip-172-31-1-193 ~]$ ls -lad /sys/fs/cgroup/*cpu*
lrwxrwxrwx. 1 root root 11 Mar  4 16:41 /sys/fs/cgroup/cpu -> cpu,cpuacct
lrwxrwxrwx. 1 root root 11 Mar  4 16:41 /sys/fs/cgroup/cpuacct -> cpu,cpuacct
drwxr-xr-x. 5 root root  0 Mar  4 16:41 /sys/fs/cgroup/cpu,cpuacct
drwxr-xr-x. 3 root root  0 Mar  4 16:41 /sys/fs/cgroup/cpuset

[centos@ip-172-31-1-193 ~]$ journalctl -u rke2-server |grep cpu
Mar 04 16:41:56 ip-172-31-1-193.us-east-2.compute.internal rke2[1033]: time="2021-03-04T16:41:56Z" level=warning msg="Disabling CPU quotas due to missing cpu.cfs_period_us"

[centos@ip-172-31-1-193 ~]$ k top pods cpulimit 
NAME       CPU(cores)   MEMORY(bytes)
cpu-demo   1962m        0Mi

```
